Blog post for API conversion (#232)
strimzi/strimzi.github.io: _posts/2021-04-18-api-conversion.md

---
layout: post
title:  "Path to CRD v1"
date: 2021-04-18
author: jakub_stejskal
---
In Strimzi 0.22 we introduced the `v1beta2` API version to prepare Strimzi for Kubernetes 1.22. The `v1beta2` API version replaces the `v1beta1` and `v1alpha1` API versions.
This post shows you how to perform an API conversion to `v1beta2` to be ready for the next Strimzi releases.

<!--more-->
  
### New API versions

> By introducing API version `v1beta2` for Strimzi resources, current Strimzi users must migrate from the old `v1beta1`  and `v1beta2` versions.

Reviewer: > the old `v1beta1` and `v1beta2` versions
this makes an implication that `v1beta2` is an old version. I think you should rephrase this sentence.

Author: ye, Paolo is right. here should be `v1alpha1` and `v1beta1`

Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources.
CRDs are themselves resources within Kubernetes.
They are installed in a Kubernetes cluster to define the versions of API for the custom resource.
Each version of the custom resource API can define its own schema for that version.
Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

In 0.22, we still support custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser`, and `KafkaTopic`.
However, these versions are currently deprecated and will be dropped in the Strimzi 0.23 release.
That's one of the reasons why users have to finish the migration before 0.23.

For the new API version, deprecated custom resource properties have been removed. You need to migrate custom resources to the new structure.
For instance, `v1beta2` only supports a new array-based listener configuration and new metrics configuration formats from config maps.

> `metrics` and `logging` properties have also changed.

Reviewer: You have already mentioned metrics change in the first sentence.

> For metrics configuration, you now use `metricsConfig` in `spec.kafka` to reference the name ofa config map with metrics configuration using the `valueFrom` property.

Reviewer: > ofa

typo

Same for `logging`, but property name remains same.
In addition, some old unused fields like `tlsSidecar` were removed as well.
All the changes to properties are fully described in our [documentation](https://strimzi.io/docs/operators/latest/full/deploying.html#proc-upgrade-cli-tool-files-str).

> You can also see the deprecated property as a warning in the custom resource status

Reviewer: > You can also see the deprecated property as a warning in the custom resource status

This feels weird. I would say `You can also see a warning in the custom resource status when the deprecated property is used.` or similar.

For smooth migration to new custom resources and CRDs version, we provide an `api-conversion` tool. _Note that for proper conversion you need to have Strimzi 0.22 installed with all its CRDs._

### API conversion tool

The `api conversion` tool is shipped with Strimzi 0.22 as a [zip file](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.zip) and [tar file](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.tar.gz).

> After extracting, you will find all the libs needed by the tool in `libs` directory and two scripts for running the tool in the folder `bin`.

Reviewer: directory/folder

The output is printed to _stdout_, or you can specify an output file with `-o` or `--output` options.
You can also use the `--in-file` option to save changes directly to input file.
For more information, run the `bin/api-conversion.sh cf --help` command.

#### convert-resources

The `convert-resources` (`cr`) command converts custom resources directly in your Kubernetes cluster.
You can use `-a`,`--all-namespaces` or `-n`,`--namespace` options to specify namespace where the resources should be converted.
With `-k`,`--kind` the tool will convert only a specific kind of custom resources.
Or the command can convert multiple kinds.
For instance, to convert `Kafka` and `KafkaTopic` you just need to pass `--kind Kafka --kind KafkaTopic` to the `cr` command.

You can use the `--name` option to specify a custom resource by name.
This option can only be used with `--namespace` and single `--kind` options.
For more information, you can run the `bin/api-conversion.sh cr --help` command.

> 

Reviewer: Do you think you can put there an example command?

#### crd-upgrade

The `crd-upgrade` (`crd`) command updates `spec.versions` in the Strimzi CRDs to declare `v1beta2` as the storage API version.
The command also updates the API version of existing custom resources and stores them as `v1beta2`.

Once you upgrade CRDs, you must use only the `v1beta2` version in Strimzi Custom resources.

> 

Reviewer: Do you think you can put there an example command?

---
layout: post
title:  "Path to CRD v1"
date: 2021-04-18
author: jakub_stejskal
---

> In Strimzi 0.22 we introduced new API version `v1beta2` which will replace `v1beta1` / `v1alpha1` APIs and make Strimzi ready for Kubernetes 1.22.

Reviewer 2: ```suggestion
In Strimzi 0.22 we introduced the `v1beta2` API version to prepare Strimzi for Kubernetes 1.22. The `v1beta2` API version replaces the `v1beta1` and `v1alpha1` API versions .
```

> This blog post will show you how to perform the conversion and get ready for next Strimzi releases.

Reviewer 2: ```suggestion
This post will shows you how to perform an API conversion to `v1beta2` to be ready for the next Strimzi releases.
```

<!--more-->
  
### New API versions

> As we stated before, we introduced new API version `v1beta2` for Strimzi resources.
> Every user of Strimzi will have to migrate from old `v1beta1` / `v1beta2`.

Reviewer 2: ```suggestion
By introducing API version `v1beta2` for Strimzi resources, all Strimzi users must migrate from the old `v1beta1`  and `v1beta2` versions.
```

This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23 or later.

> The main reason why Strimzi users had to do this migration are preparations of Strimzi to use [Kubernetes CRD `apiextensions/v1`](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122).

Reviewer 2: ```suggestion
The main reason for the migration is to prepare Strimzi to use [Kubernetes CRD `apiextensions/v1`](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122).
```

Custom resources are created and controlled using APIs added to Kubernetes by CRDs.
Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources.
CRDs are themselves resources within Kubernetes.
They are installed in a Kubernetes cluster to define the versions of API for the custom resource.
Each version of the custom resource API can define its own schema for that version.
Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

> In 0.22, we still support custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.

Reviewer 2: ```suggestion
In 0.22, we still support custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser`, and `KafkaTopic`.
```

> However, these versions are currently deprecated and will be dropped in the following Strimzi 0.23.

Reviewer 2: ```suggestion
However, these versions are currently deprecated and will be dropped in the Strimzi 0.23 release.
```

> That's one of the reasons, why users have to finish the migration before 0.23.

Reviewer 2: ```suggestion
That's one of the reasons why users have to finish the migration before 0.23.
```

> One another change is, that with new API version we remove deprecated fields so users need to migrate them to new format.

Reviewer 2: ```suggestion
For the new API version, deprecated custom resource properties have been removed. You need to migrate custom resources to the new structure.
```

> For instance `v1beta2` supports only new array based listener configuration and new metrics configuration formats from config maps.

Reviewer 2: ```suggestion
For instance `v1beta2` only supports a new array-based listener configuration and new metrics configuration formats from config maps.
```

> `metrics` and `logging` field were also changed as well.
> For metrics configuration you should use `metricsConfig` in `spec.kafka` to reference name of config map with metrics configuration in `valueFrom` field.
> Same for `logging`, but field name remains same.
> In addition, some old unused fields like `tlsSidecar` were removed as well.
> All changes in fields are described in our [documentation](https://strimzi.io/docs/operators/latest/full/deploying.html#proc-upgrade-cli-tool-files-str).
> You can also see the deprecated field in CR status as warnings.

Reviewer 2: ```suggestion
`metrics` and `logging` properties have also changed.
For metrics configuration, you now use `metricsConfig` in `spec.kafka` to reference the name ofa config map with metrics configuration using the `valueFrom` property.
Same for `logging`, but property name remains same.
In addition, some old unused fields like `tlsSidecar` were removed as well.
All the changes to properties are fully described in our [documentation](https://strimzi.io/docs/operators/latest/full/deploying.html#proc-upgrade-cli-tool-files-str).
You can also see the deprecated property as a warning in the custom resource status.
```

> For smooth migration to new CRs and CRDs version we created `api-conversion` tool. _Note that for proper conversion you need to have installed Strimzi 0.22 with all its CRDs._

Reviewer 2: ```suggestion
For smooth migration to new custom resources and CRDs version, we provide an `api-conversion` tool. _Note that for proper conversion you need to have Strimzi 0.22 installed with all its CRDs._
```

### API conversion tool

> The tool is shipped as part of Strimzi 0.22 in [zip](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.zip) and [tar.gz](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.tar.gz) archives.
> After un-archiving, you will find there all libs needed by the tool in `libs` directory and two scripts for running the tool in folder `bin`.
> Script `api-conversion.sh` is for linux/mac users and `api-conversion.cmd` is for Windows users.

Reviewer 2: ```suggestion
The `api conversion` tool is shipped with Strimzi 0.22 as a [zip file](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.zip) and [tar file](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.tar.gz).
After extracting, you will find all the libs needed by the tool in `libs` directory and two scripts for running the tool in the folder `bin`.
The `api-conversion.sh` script is for linux/mac users and `api-conversion.cmd` is for Windows users.
```

> The tool can operate in two modes:

Reviewer 2: ```suggestion
The tool operates in two modes:
```

> * convert your resources to `v1beta2`. It can convert resources in yaml files or resource already deployed into Kubernetes cluster.

Reviewer 2: ```suggestion
* Converting custom resources to `v1beta2`. It can convert resources in yaml files or resource already deployed into Kubernetes cluster.
```

> * convert CRDs to use `v1beta2` as a stored version and use `v1beta2` in all Strimzi custom resources.

Reviewer 2: ```suggestion
* Converting CRDs to use `v1beta2` as a stored version and use `v1beta2` in all Strimzi custom resources.
```

> You can use the tool directly from un-archived directory as it's shown in the example bellow.

Reviewer 2: ```suggestion
You can use the tool directly from an extracted directory as shown in the example below.
```

![api-conversion tool help](/assets/images/posts/2021-04-18-api-conversion.png)

#### convert-file

> `convert-file` (`cf`) command basically takes input yaml file passed via option `-f` or `--file` and convert CR to new API version and migrate deprecated fields.

Reviewer 2: ```suggestion
The `convert-file` (`cf`) command updates the YAML file of a custom resource, which is passed using an `-f` or `--file` parameter. It converts the custom resource to new API version and updates the properties into the current structure.
```

> The output will be printed to _stdout_ or you can specify output file with `-o` or `--output` options.
> You can also use `--in-file` option to save changes directly to input file.
> For more info run `bin/api-conversion.sh cf --help` command.

Reviewer 2: ```suggestion
The output is printed to _stdout_, or you can specify an output file with `-o` or `--output` options.
You can also use the `--in-file` option to save changes directly to input file.
For more information, run the `bin/api-conversion.sh cf --help` command.
```

#### convert-resources

> `convert-resources` (`cr`) allow you to convert custom resources directly in your Kubernetes cluster.

Reviewer 2: ```suggestion
The `convert-resources` (`cr`) command allows you to convert custom resources directly in your Kubernetes cluster.
```

You can use `-a`,`--all-namespaces` or `-n`,`--namespace` options to specify namespace where the resources should be converted.
With `-k`,`--kind` the tool will convert only a specific kind of custom resources.

> This option can be passed multiple times to convert multiple kinds.

Reviewer 2: ```suggestion
Or the command can convert multiple kinds.
```

> For instance to convert `Kafka` and `KafkaTopic` you just need to pass `--kind Kafka --kind KafkaTopic` to `cr` command.

Reviewer 2: ```suggestion
For instance, to convert `Kafka` and `KafkaTopic` you just need to pass `--kind Kafka --kind KafkaTopic` to the `cr` command.
```

> The last option is `--name` where you can specify a name of the custom resource.

Reviewer 2: ```suggestion
You can use the `--name` option to specify a custom resource by name.
```

> This option can be used only with `--namespace` and single `--kind` options.

Reviewer 2: ```suggestion
This option can only be used with `--namespace` and single `--kind` options.
```

> For more info run `bin/api-conversion.sh cr --help` command.

Reviewer 2: ```suggestion
For more information, you can run the `bin/api-conversion.sh cr --help` command.
```

#### crd-upgrade

> `crd-upgrade` (`crd`) updates `spec.versions` in the Strimzi CRDs to declare `v1beta2` as the storage API version.

Reviewer 2: ```suggestion
The `crd-upgrade` (`crd`) command updates `spec.versions` in the Strimzi CRDs to declare `v1beta2` as the storage API version.
```

> The command also updates existing custom resources where it updates API version and stores them as `v1beta2`.

Reviewer 2: ```suggestion
The command also updates the API version of existing custom resources and stores them as `v1beta2`.
```

> Once you upgrade CRDs, you should use only `v1beta2` version in Strimzi Custom resources.

Reviewer 2: ```suggestion
Once you upgrade CRDs, you must use only the `v1beta2` version in Strimzi Custom resources.
```

### Conclusion

> In this blog post we take a brief look to API conversion, which will be needed for future Strimzi versions.

Reviewer 2: ```suggestion
In this post we took a brief look at API conversion, which is needed for future Strimzi versions.
```

Each version of the custom resource API can define its own schema for that version.
Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

In 0.22, we still support custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.
However, these versions are currently deprecated and will be dropped in the following Strimzi 0.23.
That's one of the reasons, why users have to finish the migration before 0.23.

One another change is, that with new API version we remove deprecated fields so users need to migrate them to new format.
For instance `v1beta2` supports only new array based listener configuration and new metrics configuration formats from config maps.
`metrics` and `logging` field were also changed as well.
For metrics configuration you should use `metricsConfig` in `spec.kafka` to reference name of config map with metrics configuration in `valueFrom` field.
Same for `logging`, but field name remains same.

> In additional some old unused fields like `tlsSidecar` were removed as well.

Reviewer 3: ```suggestion
In addition some old unused fields like `tlsSidecar` were removed as well.
```

After un-archiving, you will find there all libs needed by the tool in `libs` directory and two scripts for running the tool in folder `bin`.
Script `api-conversion.sh` is for linux/mac users and `api-conversion.cmd` is for Windows users.

The tool can operate in two modes:
* convert your resources to `v1beta2`. It can convert resources in yaml files or resource already deployed into Kubernetes cluster.
* convert CRDs to use `v1beta2` as a stored version and use `v1beta2` in all Strimzi custom resources.

You can use the tool directly from un-archived directory as it's shown in the example bellow.

![api-conversion tool help](/assets/images/posts/2021-04-18-api-conversion.png)

#### convert-file

`convert-file` (`cf`) command basically takes input yaml file passed via option `-f` or `--file` and convert CR to new API version and migrate deprecated fields.
The output will be printed to _stdout_ or you can specify output file with `-o` or `--output` options.
You can also use `--in-file` option to save changes directly to input file.

> For more info run `bin/api-conversion.sh cr --help` command.

Reviewer 3: ```suggestion
For more info run `bin/api-conversion.sh cf --help` command.
```

#### convert-resources

`convert-resources` (`cr`) allow you to convert custom resources directly in your Kubernetes cluster.
You can use `-a`,`--all-namespaces` or `-n`,`--namespace` options to specify namespace where the resources should be converted.
With `-k`,`--kind` the tool will convert only a specific kind of custom resources.
This option can be passed multiple times to convert multiple kinds.
For instance to convert `Kafka` and `KafkaTopic` you just need to pass `--kind Kafka --kind KafkaTopic` to `cr` command.

> The last option is `-n`,`--name` where you can specify a name of the custom resource.

Reviewer 3: Do we really have `-n` for name and `-n` for namespace?

Author: `-n` is just for namespace, will fix it

This option can be used only with `--namespace` and single `--kind` options.

> For more info run `bin/api-conversion.sh cf --help` command.

Reviewer 3: ```suggestion
For more info run `bin/api-conversion.sh cr --help` command.
```

> #### convert-crd

Reviewer 3: ```suggestion
#### upgrade-crd
```

> `convert-crd` updates `spec.versions` in the Strimzi CRDs to declare `v1beta2` as the storage API version.

Reviewer 3: ```suggestion
`upgrade-crd` updates `spec.versions` in the Strimzi CRDs to declare `v1beta2` as the storage API version.
```

The command also updates existing custom resources where it updates API version and stores them as `v1beta2`.

> Once you convert CRDs to API version `apiextensions/v1`, you should use only `v1beta2` version in Strimzi Custom resources.

Reviewer 3: ```suggestion
Once you upgrade CRDs, you should use only `v1beta2` version in Strimzi Custom resources.
```

---
layout: post

> title:  "API versions conversion"

Reviewer 3: Maybe we can also come up with some better title? _Path to CRD v1_? _Changes to Strimzi API versions_?

Author: `Path to CRD v1` sounds good to me :)

date: 2021-04-18
author: jakub_stejskal
---

> Upgrade process of Strimzi and Kafka is not trivial.

Reviewer 3: I would not refer to this as the upgrade process ... I would probably remove this sentence.

Author: Ye, it's leftover

In Strimzi 0.22 we introduced new API version `v1beta2` which will replace `v1beta1` / `v1alpha1` APIs and make Strimzi ready for Kubernetes 1.22.
This blog post will show you how to perform the conversion and get ready for next Strimzi releases.

<!--more-->

Even with these changes, there are still some disadvantages:
* For upgrade, roll out the new brokers while first using the older `log.message.format.version` or `inter.broker.protocol.version` and only afterwards change to the new versions for message format and inter-broker protocol.
* Downgrade won't be executed when new `log.message.format.version` or `inter.broker.protocol.version` are already used.
  You need to change configuration to use same `log.message.format.version` and `inter.broker.protocol.version` as older Kafka.

You can find further information in our [documentation](https://strimzi.io/docs/operators/0.22.1/full/deploying.html#proc-upgrade-cli-tool-crds-str).

### New API versions

As we stated before, we introduced new API version `v1beta2` for Strimzi resources.

> Every user of Strimzi will have to the migration from old `v1beta1` to `v1beta2`.

Reviewer 3: ```suggestion
Every user of Strimzi will have to migrate from old `v1beta1` to `v1beta2`.
```

This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23 or later.
The main reason why Strimzi users had to do this migration are preparations of Strimzi to use [Kubernetes CRD `apiextensions/v1`](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122).

> Custom resources are edited and controlled using APIs added to Kubernetes by CRDs.

Reviewer 3: The _edited_ sounds a bit weird here.

Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources.
CRDs are themselves resources within Kubernetes.
They are installed in a Kubernetes cluster to define the versions of API for the custom resource.
Each version of the custom resource API can define its own schema for that version.
Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

In 0.21, we supported custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.
These versions are currently deprecated and will be dropped in the following Strimzi 0.23.

> One another change is, that with new API version we remove some deprecated fields so users need to migrate them to new format.

Reviewer 3: ```suggestion
One another change is, that with new API version we remove deprecated fields so users need to migrate them to new format.
```

For instance `v1beta2` supports only new array based listener configuration and new metrics configuration formats from config maps.

> In additional some old unused fields like `tlsSidecar` were removed as well.

Reviewer 3: might be worth mentioning also the metrics or logging changes which are more important than `tlsSidecars`

> All changes in field are described in our [documentation](https://strimzi.io/docs/operators/latest/full/deploying.html#proc-upgrade-cli-tool-files-str).

Reviewer 3: Maybe you can also mention that users can see the deprecated fields the use in the conditions in the status section?

For smooth migration to new CRs and CRDs version we created `api-conversion` tool. _Note that for proper conversion you need to have installed Strimzi 0.22 with all its CRDs._

### API conversion tool

The tool is shipped as part of Strimzi 0.22 in [zip](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.zip) and [tar.gz](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/api-conversion-0.22.1.tar.gz) archives.

> After un-archive, you will find there all libs needed by the tool in `libs` directory and two scripts for running the tool in folder `bin`.

Reviewer 3: Maybe?

```suggestion
After un-archiving, you will find there all libs needed by the tool in `libs` directory and two scripts for running the tool in folder `bin`.
```

Script `api-conversion.sh` is for linux/mac users and `api-conversion.cmd` is for Windows users.

The tool can operate in two modes:
* convert your resources to `v1beta2`. It can convert resources in yaml files or resource already deployed into Kubernetes cluster.

> * convert CRDs to use `v1beta2` as a stored version and use `apiextensions/v1`.

Reviewer 3: It converts the CRDs to use `v1beta2` as the stored version and converts all previously stored resources to be stored as `v1beta2`. It does not convert anything to `apiextensions/v1` => it is just a prerequisite for it.

---
layout: post
title:  "Path to CRD v1"
date: 2021-04-18
author: jakub_stejskal
---
In Strimzi 0.22 we introduced new API version `v1beta2` which will replace `v1beta1` / `v1alpha1` APIs and make Strimzi ready for Kubernetes 1.22.
This blog post will show you how to perform the conversion and get ready for next Strimzi releases.

<!--more-->

Even with these changes, there are still some disadvantages:
* For upgrade, roll out the new brokers while first using the older `log.message.format.version` or `inter.broker.protocol.version` and only afterwards change to the new versions for message format and inter-broker protocol.
* Downgrade won't be executed when new `log.message.format.version` or `inter.broker.protocol.version` are already used.
  You need to change configuration to use same `log.message.format.version` and `inter.broker.protocol.version` as older Kafka.

> You can find further information in our [documentation](https://strimzi.io/docs/operators/0.22.1/full/deploying.html#proc-upgrade-cli-tool-crds-str).

Reviewer 3: This should not be here, or? Apart from the link which should be maybe at the end?

Author: ye, it's leftover

### New API versions

As we stated before, we introduced new API version `v1beta2` for Strimzi resources.

> Every user of Strimzi will have to migrate from old `v1beta1` to `v1beta2`.

Reviewer 3: ```suggestion
Every user of Strimzi will have to migrate from old `v1alpha1` / `v1beta1` to `v1beta2`.
```

This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23 or later.
The main reason why Strimzi users had to do this migration are preparations of Strimzi to use [Kubernetes CRD `apiextensions/v1`](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122).

Custom resources are created and controlled using APIs added to Kubernetes by CRDs.
Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources.
CRDs are themselves resources within Kubernetes.
They are installed in a Kubernetes cluster to define the versions of API for the custom resource.

> Each version of the custom resource API can define its own schema for that version. have to the migration from old `v1bet

Reviewer 3: It looks like this is not finished.

Author: Copy/paste issue

Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version

> In 0.21, we supported custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.

Reviewer 3: We still support them in 0.22.
